# NovoBridge

This is the repository for the NovoBridge pipeline, as described in:<br>
Hugo B. C. Kleikamp, Mario Pronk, Claudia Tugui, Leonor Guedes da Silva, Ben Abbas, Yue Mei Lin, Mark C.M. van Loosdrecht and Martin Pabst* Quantitative profiling of microbial communities by de novo metaproteomics, BiorXiv (2020) (updated as soon as accepted)      

Available from:<br>
https://www.biorxiv.org/content/10.1101/2020.08.16.252924v1


#### What is Novobridge?
Novobridge is an automated pipeline for fast processing, and integrated annotation and visualization of *de novo* proteomics data.

#### How does it work?
The current version of the pipeline is included as a single script: <code>Novobridge.py</code>, which can be run using the command line or from any python interpreter. The pipeline will be turned into a module with separate functions in the near future. The pipeline uses UniPept API methods `pept2lca` and `pept2fun` to annotate taxonomy and function, and uses the KEGG database to match the functional annotations to pathways based on EC numbers.

#### What does it do?
1. Parse & pre-process peptides
2. Submit peptides to UniPept server
3. Post-process annotations
3. Taxonomic quantification and visualiztion
4. Functional annotation and quantification with KEGG orthologies

#### What input files does it use?
- NovoBridge can work with filetypes -.csv, -.xls, or in -.xlsx, which either simply contain a list of peptides (with header: Peptide),
  <br> or with file formats from PEAKs and DeepNovo. The input files need to be stored in the folder input_peaks. 
- Apart from the input files, there are two utilities files: <code>keg.pkl</code> which is required for functional annotation, and <code>Krona_template.xlsm</code>, which is required for Krona-plot visualization. Both files can be created with the script <code>download_utilities.py</code>

#### What outputs does it generate?
The outputs generated by the pipeline are distributed over 3 folders. For each file in input_peaks the following outputs are generated: 
1. output unipept: input file, annotated with unipept for each seaparate peptide.
2. output composition: quantified taxonomic distributions, krona plots and stacked bar charts.
3. output function: quantified KEGG pathways


#### Licensing

The pipeline is licensed with standard MIT-license. <br>
If you would like to use this pipeline in your research, please cite the following papers: 
      
- Hugo B. C. Kleikamp, Mario Pronk, Claudia Tugui, Leonor Guedes da Silva, Ben Abbas, Yue Mei Lin, Mark C.M. van Loosdrecht and Martin Pabst* Quantitative profiling of microbial communities by de novo metaproteomics, BiorXiv (2020) (updated as soon as accepted)           

- Robbert Gurdeep Singh, Alessandro Tanca, Antonio Palomba, Felix Van der Jeugt, Pieter Verschaffelt, Sergio Uzzau, Lennart Martens, Peter Dawyndt, and Bart Mesuere. (2019). Unipept 4.0: Functional Analysis of Metaproteome Data. J. Proteome Res. 2019, 18, 606−615 Article. 
- Kanehisa, M., & Goto, S. (2000). KEGG : Kyoto Encyclopedia of Genes and Genomes, 28(1), 27–30.




#### Contact:
-Hugo Kleimamp(Developer): H.B.C.Kleikamp@tudelft.nl<br> 
-Martin Pabst: M.Pabst@tudelft.nl<br>

#### Links to other repositories:
https://github.com/unipept<br>
https://github.com/marbl/Krona


